### ISSUE TITLE
Boundary and manifold IDs not respected on boundary

### ISSUE THREAD

**Issue by jppelteret** [dealii/dealii#5931]
There seems to be an issue when refining grids that have both manifold and boundary IDs, and one is not precise about providing a non-overlapping enumeration for all of these IDs. For example, consider a problem where one has two surfaces (one flat and one curved), and you assign a boundary ID to the one and a manifold ID to the other. If they are given the same enumeration and if a curved manifold describes the latter then you can get unexpected results. Its even more confusing if the manifold IDs are prescribed to the interior of the domain, as they have nothing to do with the boundary at all.

From the explanatory (and, hopefully, sufficiently minimal) example given below, here's the expected result 
![bid_mid_unique](https://user-images.githubusercontent.com/10460610/36428177-899d4832-164f-11e8-9e6c-25c5ef33f0b3.png)
and the actual result
![bid_mid_clash](https://user-images.githubusercontent.com/10460610/36428176-898073ce-164f-11e8-8375-61150de7c14d.png)

Although the reported boundary and manifold ID counts may seem correct
```
  Number of active cells:       128
  Boundary: 1  ;  Number of faces: 8
  Boundary: 2  ;  Number of faces: 8
  Boundary: 3  ;  Number of faces: 8
  Boundary: 4  ;  Number of faces: 8
  Boundary: 5  ;  Number of faces: 32
  Manifold: -1  ;  Number of faces: 256
  Manifold: 1  ;  Number of faces: 32
```
but the grid itself may still be distorted upon refinement.

```c++

#include <deal.II/grid/grid_generator.h>
#include <deal.II/grid/grid_tools.h>
#include <deal.II/grid/grid_out.h>
#include <deal.II/grid/manifold_lib.h>
#include <deal.II/fe/fe_q.h>
#include <deal.II/dofs/dof_handler.h>
#include <deal.II/lac/vector.h>
#include <deal.II/numerics/data_out.h>

#include <fstream>

using namespace dealii;

template <typename Stream, int dim>
void print_triangulation_data(Stream &stream, 
                              const Triangulation<dim> &triangulation)
{
 // Number of active cells on the triangulation
 stream
 << "  Number of active cells:       "
 << triangulation.n_active_cells()
 << "\n";

 // Boundary id count
 std::map<int,int> boundary_id_count;
 std::map<int,int> manifold_id_count;
 for (typename Triangulation<dim>::active_cell_iterator
     cell = triangulation.begin_active();
     cell!=triangulation.end(); ++cell)
 {
   for (unsigned int face=0; face<GeometryInfo<dim>::faces_per_cell; ++face)
   {
     if (cell->face(face)->at_boundary())
     {
       boundary_id_count[cell->face(face)->boundary_id()]++;
       manifold_id_count[cell->face(face)->manifold_id()]++;
     }
     else
     {
       // Prevent double-accounting when face is shared
       if (cell->neighbor_level(face) == cell->level())
       {
         if (cell->id() < cell->neighbor(face)->id())
           manifold_id_count[cell->face(face)->manifold_id()]++;
       }
       else
         manifold_id_count[cell->face(face)->manifold_id()]++;

     }
   }
 }

 for (typename std::map<int,int>::const_iterator
     it = boundary_id_count.begin();
     it != boundary_id_count.end(); ++it)
 {
   stream
   << "  Boundary: " << it->first
   << "  ;"
   << "  Number of faces: " << it->second
   << "\n";
 }
 for (typename std::map<int,int>::const_iterator
     it = manifold_id_count.begin();
     it != manifold_id_count.end(); ++it)
 {
   stream
   << "  Manifold: " << it->first
   << "  ;"
   << "  Number of faces: " << it->second
   << "\n";
 }

 stream << std::endl;
}

int main (int argc, char *argv[])
{
  Utilities::MPI::MPI_InitFinalize mpi_initialization(argc, argv, 1);

  const int dim = 2;
  const unsigned int n_global_refinements = 2;

  // Create a geometry with flat and curved boundaries
  const double inner_radius = 0.25;
  const double outer_radius = 0.5;
  const double tol = 1e-6;
  Triangulation<dim> tria;
  GridGenerator::hyper_cube_with_cylindrical_hole (tria,inner_radius,outer_radius);

  // Enumerate the flat boundaries and the curved one
  // separately. Also provide a manifold ID to the curved
  // surface. Note that the manifold ID coincides with
  // one of the boundary IDs. This is the root cause of the
  // issue...
  const types::manifold_id curved_manifold_id = 1; 
  // const types::manifold_id curved_manifold_id = 5; // This would work fine 

  for (typename Triangulation<dim>::active_cell_iterator
     cell = tria.begin_active();
     cell != tria.end(); ++cell)
 {
   for (unsigned int face=0; face<GeometryInfo<dim>::faces_per_cell; ++face)
   {
     if (cell->face(face)->at_boundary())
     {
      const Point<dim> pt_centre = cell->face(face)->center();

      for (unsigned int vertex=0; vertex<GeometryInfo<dim>::vertices_per_face; ++vertex)
      {
        const Point<dim> pt_vertex = cell->face(face)->vertex(vertex);

        if (std::abs(pt_vertex.norm() - inner_radius) < tol)
        {
          cell->face(face)->set_manifold_id(curved_manifold_id);
          // Lets also give this curved surface a
          // unique boundary ID, just because we can.
          cell->face(face)->set_boundary_id(5);
        }
        else 
        {
          if (std::abs(pt_centre[0] - outer_radius) < tol)
            cell->face(face)->set_boundary_id(1);
          else if (std::abs(pt_centre[0] + outer_radius) < tol)
            cell->face(face)->set_boundary_id(2);
          else if (std::abs(pt_centre[1] - outer_radius) < tol)
            cell->face(face)->set_boundary_id(3);
          else if (std::abs(pt_centre[1] + outer_radius) < tol)
            cell->face(face)->set_boundary_id(4);

            // I really want to make sure that these outer
            // boundaries are treated as flat manifolds.
            cell->face(face)->set_manifold_id(numbers::flat_manifold_id);
         }
       }
     }
   }
 }

  static const SphericalManifold<dim> sphere;
  tria.set_manifold(curved_manifold_id, sphere);
  tria.refine_global(n_global_refinements);

  std::ofstream file_gridout ("grid.vtu");
  GridOut().write_vtu(tria, file_gridout);

  // Ditto for the use of higher-order mappings...
  // All flat surfaces should be treated as flat!
  // 
  // const unsigned int order = 2;
  // FE_Q<dim> fe (order);
  // DoFHandler<dim> dof_handler (tria);
  // dof_handler.distribute_dofs(fe);

  // const std::string name = "soln";
  // Vector<double> soln (dof_handler.n_dofs());
  // std::ofstream file_dataout ("soln.vtu");
  // DataOut<dim> data_out;
  // data_out.add_data_vector(dof_handler, soln, name);
  // data_out.build_patches(order);
  // data_out.write_vtu(file_dataout);

  print_triangulation_data(std::cout, tria);

  return 0;
}

```

Note: This was tested on a branch based off of 881c7a2167d1b3c98d83a5f81a4d3a1550eb4401 (so a few days old).

**Comment by luca-heltai** [comment]
Yes, this is an issue. I think the main reason for this is that there are still parts of the library where backward compatible hacks are presents...

**Comment by luca-heltai** [comment]
I have no way of running the full test suite now. Can you check if my PR fixes your problem?

**Comment by jppelteret** [comment]
Sure, will do. Thanks very much @luca-heltai !

**Comment by drwells** [comment]
Fixed by #6233. Indeed, the code provided by @jppelteret is now the `tria_boundary_id_01` test.
